web-api/src/storage/postgres.rs

```diff
@@ -414,6 +492,88 @@ impl storage::Interaction for Storage {
 
         Ok(())
     }
+
+    #[allow(clippy::too_many_lines)]
+    async fn update_interactions<F>(
+        &self,
+        user_id: &UserId,
+        updated_document_ids: &[&DocumentId],
+        mut update_logic: F,
+    ) -> Result<(), Error>
+    where
+        F: for<'a, 'b> FnMut(InteractionUpdateContext<'a, 'b>) -> &'b PositiveCoi + Send + Sync,
+    {
+        let mut tx = self.postgres.pool.begin().await?;
+
+        Database::acquire_user_coi_lock(&mut tx, user_id).await?;
+
+        let documents = self
+            .get_by_ids_with_transaction(updated_document_ids, Some(&mut tx))
+            .await?;
+
+        let now = Utc::now();
+
+        let mut document_map = documents
+            .iter()
+            .map(|d| (&d.id, d))
+            .collect::<HashMap<_, _>>();
+
+        let mut interests = Database::get_user_interests(&mut tx, user_id).await?;
+
+        let mut category_weight_diff_map = documents
+            .iter()
+            // yes this is intended to add an entry for `None` to
+            // if it appears on a document
+            .map(|d| (&d.tags, 0i32))
+            .collect::<HashMap<_, _>>();
+        let mut updates = Vec::new();
+
+        for id in updated_document_ids {
+            if let Some(document) = document_map.get_mut(id) {
+                let category_weight_diff = category_weight_diff_map
+                    .get_mut(&document.tags)
+                    .unwrap(/*we added entries for all categories*/);
+
+                let update = update_logic(InteractionUpdateContext {
+                    document,
+                    category_weight_diff,
+                    positive_cois: &mut interests.positive,
+                })
+                .clone();
+                updates.push(update);
+            } else {
+                warn!(%id, "interacted document doesn't exist");
+            }
+        }
+
+        Database::upsert_cois(&mut tx, user_id, now, &updates).await?;
+
+        Database::upsert_interactions(
+            &mut tx,
+            user_id,
+            now,
+            document_map
+                .values()
+                .map(|d| (&d.id, UserInteractionType::Positive))
+                //needed or rust fails with a `error: higher-ranked lifetime error`
+                // this seems to beg a limitation/(bug?) in rustc not our code
```

Comment:
I'm not sure it's a bug, hence the `?`. There are some limitations to the compiler  which  do not classify as bugs because they are (internally) well known limitations for the current implementation.  Like similar to how in some languages you might have a limit to the number of function arguments.